File: Benchmarks/MergeSort/Rust/merge-sort.rs
use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
};

fn merge_sort(entries: &[String]) -> Vec<String> {
    if entries.len() <= 1 {
        return entries.to_vec();
    }

    let half = entries.len() / 2;
    let left = merge_sort(&entries[0..half]);
    let right = merge_sort(&entries[half..entries.len()]);

    merge(&left, &right)
}

fn merge(a: &[String], b: &[String]) -> Vec<String> {
    let mut merged: Vec<String> = Vec::with_capacity(a.len() + b.len());
    let mut ia = 0;
    let mut ib = 0;

    while ia < a.len() && ib < b.len() {
        if a[ia] < b[ib] {
            merged.push(a[ia].clone());
            ia += 1;
        } else {
            merged.push(b[ib].clone());
            ib += 1;
        }
    }

    for i in ia..a.len() {
        merged.push(a[i].clone());
    }
    for i in ib..b.len() {
        merged.push(b[i].clone());
    }

    merged
}

fn main() {
    let args: Vec<String> = std::env::args().collect();

    let input = args.get(1).unwrap().to_string();
    let output: String = args.get(2).unwrap().to_string();

    let file = File::open(input).expect("no such file");
    let buf = BufReader::new(file);
    let lines: Vec<String> = buf
        .lines()
        .map(|l| l.expect("Could not parse line"))
        .collect();

    let sorted = merge_sort(&lines);

    let mut out = BufWriter::new(File::create(&output).unwrap());
    for line in sorted {
        let _ = writeln!(out, "{}", line);
    }
}

File: Benchmarks/MergeSort/Java/MergeSort.java

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

class MergeSort {
    public static ArrayList<String> Sort(List<String> entries) {
        if (entries.size() <= 1) {
            return new ArrayList<String>(entries);
        }

        int half = entries.size() / 2;
        ArrayList<String> left = Sort(entries.subList(0, half));
        ArrayList<String> right = Sort(entries.subList(half, entries.size()));

        return Merge(left, right);
    }

    public static ArrayList<String> Merge(List<String> a, List<String> b) {
        ArrayList<String> merged = new ArrayList<>();
        int ia = 0;
        int ib = 0;

        while (ia < a.size() && ib < b.size()) {
            if (a.get(ia).compareTo(b.get(ib)) < 0) {
                merged.add(a.get(ia));
                ia++;
            } else {
                merged.add(b.get(ib));
                ib++;
            }
        }

        merged.addAll(a.subList(ia, a.size()));
        merged.addAll(b.subList(ib, b.size()));

        return merged;
    }

    public static void main(String[] args) throws IOException {
        if (args.length < 2) {
            System.err.println("Invalid arguments.");
            System.err.println("Expected: java [file in] [file out]");
            return;
        }

        String input = args[0];
        String output = args[1];

        List<String> lines = Files.readAllLines(new File(input).toPath());
        List<String> sorted = Sort(lines);

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(output))) {
            for (String string : sorted) {
                writer.write(string);
                writer.newLine();
            }
        }
    }
}
